flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/utils/AsyncUtil.java

```diff
@@ -0,0 +1,216 @@
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            Options that = (Options) o;
+            return asyncBufferCapacity == that.asyncBufferCapacity
+                    && asyncTimeout == that.asyncTimeout
+                    && asyncOutputMode == that.asyncOutputMode;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(asyncBufferCapacity, asyncTimeout, asyncOutputMode);
+        }
+
+        @Override
+        public String toString() {
+            return asyncOutputMode + ", " + asyncTimeout + "ms, " + asyncBufferCapacity;
+        }
+    }
+
+    private static AsyncRetryStrategy<RowData> getResultRetryStrategy(
+            ExecutionConfigOptions.RetryStrategy retryStrategy,
+            Duration retryDelay,
+            int retryMaxAttempts) {
+        // Only fixed delay is allowed at the moment, so just ignore the config.
+        return new AsyncRetryStrategies.FixedDelayRetryStrategyBuilder<RowData>(
+                        retryMaxAttempts, retryDelay.toMillis())
+                .ifResult(EMPTY_RESPONSE)
+                .ifException(ANY_EXCEPTION)
+                .build();
+        //        return NO_RETRY_STRATEGY;
+    }
+
+    private static class FunctionFinder extends RexDefaultVisitor<Boolean> {
+
+        private final boolean findAsyncCall;
+        private final boolean recursive;
+
+        public FunctionFinder(boolean findAsyncCall, boolean recursive) {
+            this.findAsyncCall = findAsyncCall;
+            this.recursive = recursive;
+        }
+
+        @Override
+        public Boolean visitNode(RexNode rexNode) {
+            return false;
+        }
+
+        private boolean isImmediateAsyncCall(RexCall call) {
+            if (!(call.getOperator() instanceof BridgingSqlFunction)) {
+                return false;
+            }
+            FunctionDefinition definition =
+                    ((BridgingSqlFunction) call.getOperator()).getDefinition();
+            return AsyncScalarFunction.class.isAssignableFrom(definition.getClass());
+        }
+
+        @Override
+        public Boolean visitCall(RexCall call) {
+            boolean isImmediateAsyncCall = isImmediateAsyncCall(call);
+            return findAsyncCall == isImmediateAsyncCall
+                    || (recursive
+                            && call.getOperands().stream().anyMatch(node -> node.accept(this)));
+        }
+    }
+
+    private static class EmptyResponseResultStrategy
+            implements Predicate<Collection<RowData>>, Serializable {
+        private static final long serialVersionUID = 5065281655787318565L;
+
+        @Override
+        public boolean test(Collection<RowData> c) {
+            return null == c || c.isEmpty();
+        }
+    }
+
+    private static class AnyExceptionStrategy implements Predicate<Throwable>, Serializable {
```

Comment:
this and EmptyResponseResultStrategy seem to be runtime code and should into `flink-table-runtime`. basically all classes that a TaskManager need should go there ideally.